controllers/serviceinstance_controller.go

```diff
@@ -471,3 +605,77 @@ func (r *ServiceInstanceReconciler) getInstanceForRecovery(ctx context.Context,
 	log.Info("instance not found in SM")
 	return nil, nil
 }
+
+func (r *ServiceInstanceReconciler) shareInstance(ctx context.Context, serviceInstance *servicesv1.ServiceInstance, smClient sm.Client) error {
+	log := GetLogger(ctx)
+	var body []byte
+	share := true
+
+	if serviceInstance.Status.Shared == metav1.ConditionTrue {
+		log.Info("Service instance is shared, handling un-sharing the instance")
+		share = false
+	} else {
+		log.Info("Service instance is un-shared, handling sharing the instance")
+	}
+
+	response, err := smClient.ShareInstance(share, serviceInstance.Status.InstanceID, buildUserInfo(ctx, serviceInstance.Spec.UserInfo))
+	if err != nil {
+		log.Info(fmt.Sprintf("Got error when trying to handle instance share change: %s", err.Error()))
+		return err
+	}
+
+	if response.StatusCode == 200 {
+		return nil
+	}
+
+	body, err = bodyToBytes(response)
+	if err != nil {
+		return fmt.Errorf(fmt.Sprintf("Got error when trying to parse response body, %s", err.Error()))
+	}
+
+	return fmt.Errorf(fmt.Sprintf("Got status: %s, and response response body: %s", response.Status, body))
+}
+
+func bodyToBytes(response *http.Response) ([]byte, error) {
+	if response.Body == nil {
+		return nil, fmt.Errorf("no body in response")
+	}
+
+	defer response.Body.Close()
+
+	body, err := io.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+	return body, nil
+}
+
+func IsSharingNeedsToBeDone(conditions []metav1.Condition) bool {
+	conditionType := api.ConditionSharing
+	status := metav1.ConditionFalse
+	reason := getConditionReason(Updated, smClientTypes.PENDING)
+
+	for _, condition := range conditions {
+		if condition.Type == conditionType {
+			if strings.EqualFold(condition.Reason, reason) && condition.Status == status {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func isShareFailed(conditions []metav1.Condition) bool {
```

Comment:
duplicate code of IsSharingNeedsToBeDone (the only diff is the smClientTypes.FAILED/smClientTypes.PENDING)

Reply:
handled